modules/javafx.graphics/src/main/native-glass/win/RoActivationSupport.cpp

```diff
@@ -0,0 +1,220 @@
+        if (fnptr == nullptr) {
+            fprintf(stderr, "GetProcAddress: %s not loaded\n", name);
+            initialized = false;
+            return false;
+        }
+
+        return true;
+    }
+}
+
+void tryInitializeRoActivationSupport()
+{
+    if (initialized) {
+        return;
+    }
+
+    wchar_t path[MAX_PATH];
+    wchar_t file[MAX_PATH];
+
+    if (GetSystemDirectory(path, sizeof(path) / sizeof(wchar_t)) == 0) {
+        return;
+    }
+
+    memcpy_s(file, sizeof(file), path, sizeof(path));
+    wcscat_s(file, MAX_PATH-1, L"\\combase.dll");
+    hLibComBase = LoadLibraryW(file);
+    if (!hLibComBase) {
+        fprintf(stderr, moduleNotFoundMessage, "combase.dll");
+        return;
+    }
+
+    bool loaded =
+        loadFunction(hLibComBase, pRoInitialize, "RoInitialize") &&
+        loadFunction(hLibComBase, pRoUninitialize, "RoUninitialize") &&
+        loadFunction(hLibComBase, pRoActivateInstance, "RoActivateInstance") &&
+        loadFunction(hLibComBase, pWindowsCreateString, "WindowsCreateString") &&
+        loadFunction(hLibComBase, pWindowsDeleteString, "WindowsDeleteString");
+
+    if (!loaded) {
+        uninitializeRoActivationSupport();
+    } else {
+        HRESULT res = RoInitialize(RO_INIT_SINGLETHREADED);
+        if (FAILED(res)) {
+            fprintf(stderr, RoException("RoInitialize failed: ", res).message());
+            uninitializeRoActivationSupport();
+        } else {
+            initialized = true;
+        }
+    }
+}
+
+void uninitializeRoActivationSupport()
+{
+    if (initialized) {
+        RoUninitialize();
+    }
+
+    initialized = false;
+
+    if (hLibComBase) {
+        FreeLibrary(hLibComBase);
+        hLibComBase = NULL;
+        pRoInitialize = NULL;
+        pRoUninitialize = NULL;
+        pRoActivateInstance = NULL;
+        pWindowsCreateString = NULL;
+        pWindowsDeleteString = NULL;
+    }
+}
+
+bool isRoActivationSupported()
+{
+    return initialized;
+}
+
+HRESULT WINAPI RoInitialize(RO_INIT_TYPE initType)
+{
+    return pRoInitialize(initType);
+}
+
+void WINAPI RoUninitialize()
+{
+    pRoUninitialize();
+}
+
+HRESULT WINAPI RoActivateInstance(HSTRING activatableClassId, IInspectable** instance)
+{
+    return pRoActivateInstance(activatableClassId, instance);
+}
+
+HRESULT WINAPI WindowsCreateString(PCNZWCH sourceString, UINT32 length, HSTRING* string)
+{
+    return pWindowsCreateString(sourceString, length, string);
+}
+
+HRESULT WINAPI WindowsDeleteString(HSTRING string)
+{
+    return pWindowsDeleteString(string);
+}
+
+hstring::hstring(const char* str)
+{
+    int wstr_len = MultiByteToWideChar(CP_UTF8, 0, str, -1, nullptr, 0);
+    WCHAR* wstr = new WCHAR[wstr_len];
+    memset(wstr, 0, wstr_len * sizeof(WCHAR));
+    MultiByteToWideChar(CP_UTF8, 0, str, -1, wstr, wstr_len);
+    WindowsCreateString(wstr, wstr_len - 1, &hstr_);
+    delete[] wstr;
+}
+
+hstring::~hstring()
+{
+    WindowsDeleteString(hstr_);
```

Comment:
After adding the check for a failure to allocate the string, a null check is needed here (unless `WindowsDeleteString` specifies that it does nothing if the string is null).

Reply:
`WindowsDeleteString` does nothing when null is passed into it, but I added a null check anyway so people don't stumble upon it and wonder.